### PR TITLE
Fixes #29221 - Subnet is blank after provisioning a discovered host

### PR DESCRIPTION
**Pull request by lzap**
Subnet is blindly taken from hostgroup even if its set to nil then. The expected behavior is that subnet where discovered host was detected in should carry over to provisioning (if its nil in hostgroup).

Edit: This is no longer a doco PR, @melcorr nevermind.

### CODE REVIEW

**Comment by melcorr** [comment]
You pushed while I was reviewing it @lzap - no worries!

**Comment by lzap** [comment]
> You pushed while I was reviewing it @lzap - no worries!

Yeah because I've changed my mind, it was surprisingly easy to fix this.

**Comment by lzap** [comment]
@aruzicka I consider this an important one to get through :-)

**Comment by lzap** [comment]
Was wondering if we can squeeze this into the next release for 2.1 @adamruzicka.

File: extra/discover-host [hunk 1]
```diff
@@ -76,15 +76,21 @@ json["discovery_version"] = version
...
+    else
+      random_ip = IPAddr.new(subnet, Socket::AF_INET6) | IPAddr.new(rand(2**prefix.to_i),Socket::AF_INET6)
+      fact_name = "ipaddress6"
+    end
+    json["#{fact_name}_#{name}"] = explicit_ip ? explicit_ip : random_ip
     if name == primary
-      json["macaddress"] = mac
-      json["ipaddress"] = ip
+      json["macaddress"] = json["macaddress_#{name}"]
+      json["#{fact_name}"] = json["#{fact_name}_#{name}"]
```
Comment (adamruzicka) [inline extra/discover-host]:
```suggestion
      json[fact_name] = json["#{fact_name}_#{name}"]
```

File: extra/discover-host [hunk 2]
```diff
@@ -76,15 +76,21 @@ json["discovery_version"] = version
...
-    ipo ||= rand(1..253)
-    ip = "192.168.#{subnet}.#{ipo}"
     json["macaddress_#{name}"] = mac
-    json["ipaddress_#{name}"] = ip
+    subnet_woprefix, prefix = subnet.split('/')
+    if subnet.include?('.')
+      random_ip = IPAddr.new(subnet, Socket::AF_INET) | IPAddr.new(rand(2**prefix.to_i),Socket::AF_INET)
+      fact_name = "ipaddress"
+    else
+      random_ip = IPAddr.new(subnet, Socket::AF_INET6) | IPAddr.new(rand(2**prefix.to_i),Socket::AF_INET6)
```
Comment (adamruzicka) [inline extra/discover-host]: Nitpick `s/,S/, S/`

File: extra/discover-host [hunk 3]
```diff
@@ -76,15 +76,21 @@ json["discovery_version"] = version
...
+      fact_name = "ipaddress6"
+    end
+    json["#{fact_name}_#{name}"] = explicit_ip ? explicit_ip : random_ip
```
Comment (adamruzicka) [inline extra/discover-host:90]: How about generating a random ip on demand when it is actually needed?

**Review by adamruzicka — COMMENTED** [review]
Left a couple of comments inline. This seems like a reasonable fix to have in 2.1

File: extra/discover-host (hunk 2 above)
Comment (lzap) [reply extra/discover-host]: What's that?

Comment (lzap) [reply extra/discover-host]: Oh I get that.

File: extra/discover-host (hunk 3 above)
Comment (lzap) [reply extra/discover-host:90]: Good idea, but this is a helper script we only use when generating fake hosts during development. Do you insist? :-)

**Comment by lzap** [comment]
I have rebased and incorporated your comments. I have already branched off 16.1 unfortunately, nvm keep reviewing.

File: extra/discover-host (hunk 3 above)
Comment (adamruzicka) [reply extra/discover-host:90]: Oh, I missed it is a development only thing, guess I can let it slide then :)

File: app/controllers/discovered_hosts_controller.rb [hunk 4]
```diff
@@ -199,15 +199,17 @@ def submit_multiple_auto_provision
 
   def setup_host_class_variables
     if @host.hostgroup
+      subnet = @host.hostgroup.subnet ? @host.hostgroup.subnet : @host.subnet
```
Comment (adamruzicka) [inline app/controllers/discovered_hosts_controller.rb]: Couldn't we just do `@host.hostgroup.subnet || @host.subnet`?

Comment (timogoebel) [reply app/controllers/discovered_hosts_controller.rb]: Yep, and we can use @subnet below and do not need the helper vars. That makes this a little cleaner.

Comment (lzap) [reply app/controllers/discovered_hosts_controller.rb]: Rebasing, I am leaving the helper vars as they are used down below.

**Comment by lzap** [comment]
Rebased.

**Comment by timogoebel** [comment]
Tests are failing (though the failures look unrelated).

**Comment by lzap** [comment]
Pushed, develop is green now.

**Review by adamruzicka — APPROVED** [review]
LGTM

**Comment by timogoebel** [comment]
Thanks, @lzap.

**Pull request merged**